### ISSUE TITLE
Named query returning incorrect type list of objects with Panache

### ISSUE THREAD

**Issue by ccarv** [quarkusio/quarkus#17499]
## Describe the bug
I'm using Quarkus and Hibernate / Panache. 

For this example, I have 3 tables (`table_a`, `table_b`, `table_c`) that I am joining together using a native query. `table_b` is purely a mapping / join table for `table_a` and `table_c`:

```sql
SELECT 
  a.id,
  a.name, 
  c.loginDate 
FROM 
  TableA a 
  JOIN TableA b ON b.aId = a.id 
  JOIN TableB c ON b.cId = c.id 
WHERE 
  c.loginDate > '01-MAY-21'
```

I'm only looking for `name` and `login_date`. There is a bunch of other information stored in `table_a` and `table_c` that I don't want for this specific query. So I created an entity for this call:

```java
@Entity
@IdClass(LoginDetailsPk.class)
@NamedQuery(
    name = "LoginDetails.findFromDate",
    query = "FROM TableA a " +
            "JOIN TableA b ON b.aId = a.id " +
            "JOIN TableB c ON b.cId = c.id " +
            "WHERE c.loginDate > '01-MAY-21'"
)
public class LoginDetails extends PanacheEntityBase {

    @Id
    private int id;

    @Id
    private String name;

    @Id
    private String loginDate;

    public static List<LoginDetails> findFromDate(String fromDate) {
        // Eventually pass fromDate into find()
        return find("#LoginDetails.findFromDate").list();
    }

}
```

The following throws a class cast exception:
```java
List<LoginDetails> details = LoginDetails.findFromDate(null);

for(LoginDetails detail : details) { // <------ Throws a class cast exception
   //...
}
```

After digging into it, I'm noticing that `LoginDetails.findFromDate(...)` is not returning `List<LoginDetails>`, and is rather returning a `List<Object[]>`. The content of this list is an array of my entities, `TableA`, `TableB`, and `TableC`. 

### Expected behavior
`LoginDetails.findFromDate(...)` should return a list of `LoginDetails` and not a list of `Object[]`.

### Actual behavior
`LoginDetails.findFromDate(...)` returns an array list of `Object[]` with entites:

```
[0] -> Object [0] : TableA
[0] -> Object [1] : TableB
[0] -> Object [2] : TableC
[1] -> Object [0] : TableA
[1] -> Object [1] : TableB
[1] -> Object [2] : TableC
[2] -> Object [0] : TableA
[2] -> Object [1] : TableB
[2] -> Object [2] : TableC
```
## Environment (please complete the following information):
Windows 10
Java 1.8
Quarkus 1.13.4 Final
Maven 3.6.3

**Comment by loicmathieu** [comment]
Making a query via Hibernate with Panache should use the entity as a return type. So you need to write a named query that return a `LoginDetails` (a JPQL query).
Here you use a native Query, so Hibernate will return an `Object[]` to map each raw.

You have multiple possibilities:
- Change your request to a JPQL request returning a `LoginDetails`
- Use the `EntityManager` directly to make the query

Currently, PanacheQuery didn't support changing the bound of the return type to something that didn't extends your entity.

**Comment by ccarv** [comment]
@loicmathieu The query mentioned above is not a native query. `TableA` is a Java entity that is associated with the table `table_a`. Same applies to `TableB` and `TableC` Java entities.

A native query would look like:
```sql
SELECT 
  a.id,
  a.name, 
  c.login_date 
FROM 
  table_a a 
  JOIN table_b b ON b.a_id = a.id 
  JOIN table_c c ON b.c_id = c.id 
WHERE 
  c.login_date > '01-MAY-21'
```

**Comment by loicmathieu** [comment]
Right, I look at it took quickly, it's not a native query. 
But it's not a query that returns a `LoginDetails` neither, so Hibernate return an `Object[]` for each row.

**Comment by ccarv** [comment]
@loicmathieu I assumed that the annotation associated with the class would be it's return type. Is there a way to tell Hibernate that I'm expecting my NamedQuery to return a list of `LoginDetails`? I know I can probably do this with the `EntityManager` and a `TypedQuery` but would rather bundle the query with the class itself.

**Comment by loicmathieu** [comment]
Unfortunatly there didn't seems to be a way to do what you want.
Again, Hibernate return an `Object[]` because instead of selecting a `LoginDetails` you select multiple column so Hibernate has no way to know that you want a `LoginDetails`  object.

**Comment by FroMage** [comment]
I don't think you should create an entity class for something which isn't an entity, but you should be able to get your projection with this:

```java
@NamedQuery(
    name = "LoginDetails.findFromDate",
    query = "SELECT new my.pkg.LoginDetails(" +
            "a.id," +
            "a.name, " +
            "c.loginDate) FROM TableA a " +
            "JOIN TableA b ON b.aId = a.id " +
            "JOIN TableB c ON b.cId = c.id " +
            "WHERE c.loginDate > '01-MAY-21'"
)
public class LoginDetails {

    public final int id;
    public final String name;
    public final String loginDate;

   public LoginDetails(int id, String name, String loginDate){
      this.id = id;
      this.name = name;
      this.loginDate = loginDate;
   }

    public static List<LoginDetails> findFromDate(String fromDate) {
        // Eventually pass fromDate into find()
        return TableA.find("#LoginDetails.findFromDate").list();
    }

}

```

Something like that ought to work. See https://docs.jboss.org/hibernate/orm/5.5/userguide/html_single/Hibernate_User_Guide.html#hql-select-clause
